src/main/cpp/subsystems/Drivetrain.cpp

```diff
@@ -146,12 +148,33 @@ void Drivetrain::ControllerPeriodic() {
 
     UpdateDt();
 
-    Eigen::Vector<double, 5> y{
-        frc::AngleModulus(GetAngle()).value(), GetLeftPosition().value(),
-        GetRightPosition().value(), GetAccelerationX().value(),
-        GetAccelerationY().value()};
-    m_observer.Update(frc::Rotation2d(frc::AngleModulus(GetAngle())),
-                      GetLeftPosition(), GetRightPosition());
+    m_velObserver.Predict(m_u, GetDt());
+
+    m_leftPos = GetLeftPosition();
+    m_rightPos = GetRightPosition();
+    m_time = frc::Timer::GetFPGATimestamp();
+
+    // WPILib uses the time between pulses in GetRate() to calculate velocity,
+    // but this is very noisy for high-resolution encoders. Instead, we
+    // calculate a velocity from the change in position over change in time,
+    // which is more precise.
+    auto rawLeftVelocity = (m_leftPos - m_lastLeftPos) / (m_time - m_lastTime);
+    auto rawRightVelocity =
+        (m_rightPos - m_lastRightPos) / (m_time - m_lastTime);
+
+    m_leftVelocity = m_velocityFilter.Calculate(rawLeftVelocity);
+    m_rightVelocity = m_velocityFilter.Calculate(rawRightVelocity);
```

Comment:
You can't use the same filter on both. They can be different velocities when the robot isn't going straight, which will affect the memory of the filter.